src/Compilers/CSharp/Portable/Binder/SwitchBinder.cs

```diff
@@ -562,41 +548,36 @@ private BoundSwitchLabel BindSwitchSectionLabel(SwitchLabelSyntax node, Binder s
+                    // Bind the label case expression
+                    boundLabelExpressionOpt = sectionBinder.BindValue(caseLabelSyntax.Value, diagnostics, BindValueKind.RValue);
+                    boundLabelExpressionOpt = ConvertCaseExpression(switchGoverningType, caseLabelSyntax, boundLabelExpressionOpt, sectionBinder, ref labelExpressionConstant, diagnostics);
 
-                boundLabelExpressionOpt = ConvertCaseExpression(switchGoverningType, caseLabelSyntax, boundLabelExpressionOpt, sectionBinder, ref labelExpressionConstant, diagnostics);
-
-                // Check for bind errors
-                hasErrors = hasErrors || boundLabelExpressionOpt.HasAnyErrors;
-
-
-                // SPEC:    The constant expression of each case label must denote a value that
-                // SPEC:    is implicitly convertible (§6.1) to the governing type of the switch statement.
+                    // Check for bind errors
+                    hasErrors = hasErrors || boundLabelExpressionOpt.HasAnyErrors;
 
-                // Prevent cascading diagnostics
-                if (!hasErrors && labelExpressionConstant == null)
-                {
-                    diagnostics.Add(ErrorCode.ERR_ConstantExpected, caseLabelSyntax.Location);
-                    hasErrors = true;
-                }
+                    // SPEC:    The constant expression of each case label must denote a value that
+                    // SPEC:    is implicitly convertible (§6.1) to the governing type of the switch statement.
+                    if (!hasErrors && labelExpressionConstant == null)
+                    {
+                        diagnostics.Add(ErrorCode.ERR_ConstantExpected, caseLabelSyntax.Value.Location);
+                        hasErrors = true;
+                    }
 
-                // LabelSymbols for all the switch case labels are created by BuildLabels().
-                // Fetch the matching switch case label symbols
-                matchedLabelSymbols = FindMatchingSwitchCaseLabels(labelExpressionConstant, caseLabelSyntax);
+                    // LabelSymbols for all the switch case labels are created by BuildLabels().
+                    // Fetch the matching switch case label symbols
+                    matchedLabelSymbols = FindMatchingSwitchCaseLabels(labelExpressionConstant, caseLabelSyntax);
                     break;
                 case SyntaxKind.CasePatternSwitchLabel:
-                    // pattern matching in case is not yet implemented.
                     if (!node.HasErrors)
                     {
-                        Error(diagnostics, ErrorCode.ERR_FeatureIsUnimplemented, node, MessageID.IDS_FeaturePatternMatching.Localize());
-                        hasErrors = true;
+                        // This should not occur, because it would have been a syntax error
+                        throw ExceptionUtilities.UnexpectedValue(node.Kind());
```

Comment:
I prefer the more informative assertion.

---
In reply to: [64965549](https://github.com/dotnet/roslyn/pull/11485#discussion_r64965549) [](ancestors = 64965549)